docs/design/features/objc-interop.md

```diff
@@ -0,0 +1,164 @@
+| Objective-C object |                |     Managed proxy    |
+| Ref count: +1      |                |                      |
+|  ----------------  |                |  ------------------  |
+| | Weak reference |=| = = = = = = = >| | Strong reference | |
+| |    to proxy    | |<===============|=|    to object     | |
+|  ----------------  |                |  ------------------  |
+ --------------------                  ----------------------
+```
+
+_Note_: There is a special case with interoperability and threading in Objective-C's reference counting system - .NET [thread pools](https://docs.microsoft.com/dotnet/standard/threading/the-managed-thread-pool). The Objective-C runtime assumes each thread of execution possesses its own [`NSAutoreleasePool`](https://developer.apple.com/documentation/foundation/nsautoreleasepool) instance. In order to ensure this invariant, support for managing a thread's `NSAutoreleasePool` instance is required.
+
+**Delegates and Blocks**
+
+The expression of [Delegates][delegates_usage] in Objective-C and [Blocks][blocks_usage] in .NET represents a special challenge given the non-trivial [Block ABI][blocks_abi]. Conceptually the concept of a closure can be satisfied by using a C function with an extra "context" parameter and this exactly how one can interop with Blocks. Following the Block ABI the internal function pointer accepting the "context" can be acquire.
+
+An Objective-C Block variable defined as:
+
+```objective-c
+int(^blk)(int) = ^(int a) { return a*2; };
+```
+
+Can be queried, following the ABI, for its invoke function pointer and dispatched in C as:
+
+```C
+void* fptr = extract_using_abi(blk);
+int b = ((int(*)(id,int))fptr)(blk, a);
+```
+
+**Method dispatch**
+
+The Objective-C [message dispatch mechanism](#message_disp) will require special handling if either exception propagation or variadic variable signatures are desired. If neither exceptions nor variadic arguments are a concern then acquiring the native function pointer to the appropriate message function would be sufficient for internal uses - C# function pointers make this very easy. However, as officially described [here](https://docs.microsoft.com/xamarin/ios/internals/objective-c-selectors), users are permitted to declare their own P/Invoke signatures and perform manual message dispatch - this complicates things.
+
+Xamarin presently defines a set of hand-written assembly for the messaging functions and they handle execptions and most variadic argument cases. Then, utilizing Mono's [`mono_dllmap_insert`](http://docs.go-mono.com/?link=root:/embed) API, maps these entries to override the official Objective-C runtime APIs when loaded through a P/Invoke. This results in users only needing to know the official Objective-C runtime APIs but getting all the benefits of the hand-written assembly.
+
+There are several possible approaches to address this going forward:
+
+* Support a `mono_dllmap_insert` type API to permit overloading.
```

Comment:
With Vladimir's superhost work, this looks likely to happen. I think xam-mac should use this new hosting API instead of special-casing this in the runtime itself.

Reply:
I disagree at present. The work that @VSadov is doing is tailored at a specific scenario and I have not observed a proposal for an officially supported new API. Additionally, the need here is for a narrow subset of functions and has no need to grow. I agree that having a generic solution is appealing but I don't believe we have had a compelling user facing API proposal that would address all the complexity with replicating the `mono_dllmap_insert` API.